### PR TITLE
OTA-1338: dist/cargo_test.sh: Fix various issues 

### PR DESCRIPTION
**Pull request by DavidHurta**
The goal of this PR is to fix various issues that prevent the `cargo-test` job to run.

---

[dist/cargo_test.sh: Fail on compile errors](https://github.com/openshift/cincinnati/pull/934/commits/f0d13bce0d7a1490be2e1a6adef02e2128e6b1d7) 

Currently, the `cargo_test.sh` script skips compilation errors of tests,
subsequently does not run the non-existing tests, and exits with a zero
exit code, as no tests failed technically.

On the line:
```
mapfile -t tests < <(
    cargo test --no-run --message-format=json ${cargo_test_flags[${directory}]} | \
    jq -r 'select(.profile.test == true) | .executable'
)
```

the `cargo test` compiles tests. However, when a compilation fails, this
does not cause the script to fail. The exit code is ignored, and the
output is piped to the `jq` command, which continues with a non-zero
code as it parses the json message successfully. However, as no tests
exist, the script continues.

This commit will compile the tests and check for the exit-code of a
compilation. The `cargo test --no-run` command is used to provide
human-readable output of compilation errors. On a non-zero exit code,
the overall script fails immediately. Originally, the script failed on a
first failed test, this commit will ensure the script fails on a first
failed test or a first compilation error.

---

[dist/cargo_test.sh: Fail on any failed test](https://github.com/openshift/cincinnati/pull/934/commits/3fe2fac5e1da6b210d3a771c67c93ec43fdc6a6d) 

Fix the script to fail on any failed test. Currently, the script fails
only when the last element in the `tests` array exits with a non-zero
exit code.

Refactor the compilation of tests to fail the script using
the introduced `-e` option.

---

[prometheus-query/Cargo.toml: Add missing reqwest features](https://github.com/openshift/cincinnati/pull/934/commits/6cda3e613543f503931d643e7d429a721e1b9cd1) 

These features are needed for the code to compile.

---

[dist/cargo_test.sh: Run all tests before exiting due to failure](https://github.com/openshift/cincinnati/pull/934/commits/134b5fe87e44e9cb18388be41b53e6748fb8be76)

### CODE REVIEW

**Comment by petr-muller** [comment]
Nice find, let's check if https://github.com/openshift/release/pull/56199 fixes the compilation or if we need more (I have a feeling that we may need openssl-devel or something like that)

**Comment by DavidHurta** [comment]
We might want to fix the loop
```sh
        for test in \${tests[@]}; do
          if [[ \"${HAS_KOV}\" -eq \"1\" ]]; then
            kcov \
              --exclude-pattern=$HOME/.cargo \
              --verify \
              ${CARGO_TARGET_DIR}/cov \
              \$test
          else
            \$test
          fi
        done
```

It fails the script only when the last element of the `tests` array fails; otherwise, it passes.

```sh
$ cat dist/cargo_test.sh 
...
    (
      ${1} /usr/bin/env bash -c "\
        set -x
        cd ${directory}
        tests=(\"false\" \"false\" \"true\")    # <-------- Check here; this lines represents two failing tests and one passing
        for test in \${tests[@]}; do
          if [[ \"${HAS_KOV}\" -eq \"1\" ]]; then
            kcov \
              --exclude-pattern=$HOME/.cargo \
              --verify \
              ${CARGO_TARGET_DIR}/cov \
              \$test
          else
            \$test
          fi
        done
      "
    )
...
$ ./dist/cargo_test.sh
...
$ echo $?
0
```

**Comment by DavidHurta** [comment]
Perhaps the following code might do the trick for this PR's original goal and the https://github.com/openshift/cincinnati/pull/934#issuecomment-2326250410 comment.
```diff
         ${1} /usr/bin/env bash -c "\
-        set -x
+        set -xe
         cd ${directory}
+        cargo test --no-run
         mapfile -t tests < <(
              cargo test --no-run --message-format=json ${cargo_test_flags[${directory}]} | \
              jq -r 'select(.profile.test == true) | .executable'
         )
```

**Comment by petr-muller** [comment]
https://github.com/openshift/cincinnati/pull/935 fixed e2e :tada: 

https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cincinnati/934/pull-ci-openshift-cincinnati-master-cargo-test/1830953243715309568#1:build-log.txt%3A244 indicates we'll really need openssl-devel

**Comment by DavidHurta** [comment]
`cargo-test` is finally failing in https://github.com/openshift/release/pull/56238

**Comment by DavidHurta** [comment]
We have a few failing tests and a failing compilation upon the openssl-devel installation https://storage.googleapis.com/test-platform-results/pr-logs/pull/openshift_release/56238/rehearse-56238-pull-ci-openshift-cincinnati-master-cargo-test/1830989274061737984/build-log.txt

**Comment by DavidHurta** [comment]
/hold
The PR is still growing.

**Comment by DavidHurta** [comment]
A few failing tests. Will investigate later.
```
$ curl https://storage.googleapis.com/test-platform-results/pr-logs/pull/openshift_cincinnati/934/pull-ci-openshift-cincinnati-master-cargo-test/1831296462860849152/build-log.txt > test.log
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  991k    0  991k    0     0   684k      0 --:--:--  0:00:01 --:--:--  684k
$ cat test.log | grep "test result:"
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
test result: ok. 2 passed; 0 failed; 3 ignored; 0 measured; 0 filtered out; finished in 0.25s
test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
test result: FAILED. 13 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.06s
test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s
test result: ok. 6 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
test result: FAILED. 64 passed; 10 failed; 4 ignored; 0 measured; 0 filtered out; finished in 39.82s
test result: ok. 6 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s
test result: ok. 6 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s

```

**Comment by DavidHurta** [comment]
/unhold

All tests are able to run. We could start the review process of this PR. The failing tests would be fixed in future PRs, and existing and new PRs would at least have a functioning `cargo-test` job.

However, the latest commit 134b5fe87e44e9cb18388be41b53e6748fb8be76 seems somewhat hacky. However, straightforward and simple. I am open to suggestions. We could refactor the script, but I am cautious as to not introduce more similar issues.

**Review by petr-muller — APPROVED** [review]
/lgtm
/approve 

CI jobs are failing, which is expected now: we haven't been executing them properly, so we will merge this PR overriding the CI jobs and then submit PRs to fix the failing tests

**Comment by petr-muller** [comment]
> Not mergeable. Merge type "squash" disallowed by repo settings 

srsly?

**Review by PratikMahajan — APPROVED** [review]
/lgtm 
/retest

**Comment by petr-muller** [comment]
LOL we broke the logic again because https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cincinnati/934/pull-ci-openshift-cincinnati-master-customrust-cargo-test/1831755007058251776 PASSes despite failing to compile:

https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cincinnati/934/pull-ci-openshift-cincinnati-master-customrust-cargo-test/1831755007058251776/artifacts/cargo-test/cargo-test/build-log.txt

/hold

File: dist/cargo_test.sh [hunk 1]
```diff
@@ -31,16 +32,17 @@ function run_tests() {
...
   for directory in ${!cargo_test_flags[*]}; do
     if [[ "${HAS_KOV}" -eq "1" ]]; then
       # we want to prevent completely untested functions to be stripped
       export RUSTFLAGS='-C link-dead-code'
     fi
 
-    (
-      ${1} /usr/bin/env bash -c "\
-        set -x
+    if /usr/bin/env bash -c "\
```
Comment (petr-muller) [inline dist/cargo_test.sh]:
```suggestion
    if ! /usr/bin/env bash -c "\
```

oopsiez

Comment (petr-muller) [reply dist/cargo_test.sh]:
```
 test result: ok. 6 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
+ has_failed=true 
```
LOL https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cincinnati/934/pull-ci-openshift-cincinnati-master-cargo-test/1831755007003725824#1:build-log.txt%3A31441

Comment (DavidHurta) [reply dist/cargo_test.sh]: Damn :skull:

**Comment by DavidHurta** [comment]
No compile errors in `ci/prow/customrust-cargo-test` thanks to https://github.com/openshift/release/pull/56354:
```
$ curl https://storage.googleapis.com/test-platform-results/pr-logs/pull/openshift_cincinnati/934/pull-ci-openshift-cincinnati-master-customrust-cargo-test/1831990990823493632/build-log.txt > tmp.file && grep -ie "could not compile" tmp.file
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  992k    0  992k    0     0  5117k      0 --:--:-- --:--:-- --:--:-- 5139k
```

**Review by petr-muller — APPROVED** [review]
/lgtm
/approve
/meow
/hold cancel
/override ci/prow/cargo-test
/override ci/prow/customrust-carg-test

These jobs fail with genuine failures. These were previously just hidden, so we are merging them to fix the CI signal. It means our other PRs will fail them, but tests were failing before, we just did not know about it. We can either explicitly override, or prioritize fixing the tests.

**Comment by DavidHurta** [comment]
@petr-muller, there is a typo in your override.

/honk

**Pull request merged**